scripts/o3de/tests/test_compatibility.py

```diff
@@ -181,3 +181,47 @@ def test_resolve_gem_dependencies(gem_names, all_gem_json_data, expected_result)
     else:
         assert not expected_result
         assert errors
+
+@pytest.mark.parametrize(
+    "num_root_gems, num_dependencies_per_gem, num_optional_gems, tree_depth", [
+        # Given:
+        # R num_root_gems
+        # N num_dependencies_per_gem
+        # D tree_depth - 1
+        # Total gems = R * (pow(N, D + 1) - 1) / (N -1)
+
+        # Single gem with no dependencies, 1 gem
+        pytest.param(1, 0, 0, 1),
+        # 100 root gems with no dependencies, 100 gems
+        pytest.param(100, 0, 0, 1),
+        # 1 root gem with 3 dependencies, 2 levels, 4 gems
+        pytest.param(1, 3, 0, 2),
+        # 10 root gems with 3 dependencies, 2 levels, 40 gems
+        pytest.param(10, 3, 0, 2),
+        # 1 root gem with 3 dependencies, 3 levels, 13 gems
+        pytest.param(1, 3, 0, 3),
+        # 10 root gems with 2 dependencies, 5 levels, 310 gems
+        pytest.param(10, 2, 0, 5)
```

Comment:
Not a single one of these test, validate with any optional gem dependencies.

Reply:
good catch, will add some